apm-lambda-extension/logsapi/subscribe_test.go

```diff
@@ -0,0 +1,148 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSubscribeWithSamLocalEnv(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	os.Setenv("AWS_SAM_LOCAL", "true")
+	t.Cleanup(func() {
+		os.Unsetenv("AWS_SAM_LOCAL")
+	})
+	out := make(chan LogEvent)
+
+	err := Subscribe(ctx, "testID", []EventType{Platform}, out)
+	assert.Error(t, err)
+}
+
+func TestSubscribeAwsRequest(t *testing.T) {
+	listenerHost = "localhost"
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	out := make(chan LogEvent)
+	// For subscription request
+	expectedTypes := []EventType{Platform}
+	expectedBufferingCfg := BufferingCfg{
+		MaxItems:  10000,
+		MaxBytes:  262144,
+		TimeoutMS: 25,
+	}
+
+	// Create aws runtime API server and handler
+	awsRuntimeApiServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		request_bytes, _ := ioutil.ReadAll(r.Body)
+		req := SubscribeRequest{}
+		json.Unmarshal(request_bytes, &req)
+		// Validate the subscription request
+		assert.Equal(t, req.BufferingCfg, expectedBufferingCfg)
+		assert.Equal(t, req.EventTypes, expectedTypes)
+	}))
+	defer awsRuntimeApiServer.Close()
+
+	// Set the Runtime server address as an env variable
+	os.Setenv("AWS_LAMBDA_RUNTIME_API", awsRuntimeApiServer.Listener.Addr().String())
+
+	// Subscribe to the logs api and start the http server listening for events
+	err := Subscribe(ctx, "testID", []EventType{Platform}, out)
+	if err != nil {
+		t.Logf("Error subscribing, %v", err)
+		t.Fail()
+		return
+	}
+	defer logsAPIServer.Close()
+
+	// Create a request to send to the logs listener
+	platformDoneEvent := `{
+		"time": "2021-02-04T20:00:05.123Z",
+		"type": "platform.runtimeDone",
+		"record": {
+		   "requestId":"6f7f0961f83442118a7af6fe80b88",
+		   "status": "success"
+		}
+	}`
+	body := []byte(`[` + platformDoneEvent + `]`)
+	url := "http://" + logsAPIListener.Addr().String()
+	req, err := http.NewRequest("GET", url, bytes.NewReader(body))
+	if err != nil {
+		t.Log("Could not create request")
+	}
+
+	// Send the request to the logs listener
+	client := &http.Client{}
+	go func() {
+		_, err = client.Do(req)
```

Comment:
is there a reason this is being done in a go routine? The way this is written there's a race between `if err != nil` and `client.Do` executing and setting err

Reply:
Good point about the race. It's in a go routine so that I can read from the channel in line 108. 
I could put the check for an err in the go routine and then just receive from the `out` channel, if that would make sense?

Reply:
is the read from `out` synchronized with the client request returning? you could also make `out` a buffered channel, so it will store `event` without needing someone to read from it immediately, which would allow you to remove the go routine. There's still a race between the test function exiting and the `if err != nil`, since they'll be executed concurrently

Reply:
Good idea on making it a buffered channel. I think I can make this change in other tests as well.